Upload image menu (cover, profile)

rkzel opened the pull request:
This is looking more like in design. "Update photo" is now just a header, clickable ~(not a button though)~ to expand the menu.

rkzel committed:
image upload menu trued up

rkzel committed:
added camera icon, regulated padding

Schwartz10 commented on src/client/components/ImageMenu.js:
```diff
@@ -102,29 +106,53 @@ const ImageMenu = ({
-      {imageExists && (
-        <div onClick={() => dispatchModal(removeItem(imageCid, imageTag))}>
-          Delete
+    <ClickOutHandler onClickOut={() => toggleImageMenu('')}>
```
I agree that it seems to make sense to add a dependency for this functionality, but how did you choose this dependency specifically? Are you sure it's sustainable? I get concerned that the last time an update was made to this package was december 2018, and its react version is pre-hooks (which usually is a danger sign).

chadoh replied:
No dependency necessary! Add a wrapping div with a `tabIndex="-1"`. This will make it respond to `onFocus` and, more importantly, `onBlur` events. 

```jsx
<div tabIndex="-1" onBlur={() => toggleImageMenu('')}>
  <ImageMenuStyled ... />
</div>
```

rkzel replied:
A-ha. I took ClickOutHandler from Aragon's DropDown, and now replaced it as suggested - and it working - thank you! :)

Schwartz10 reviewed (commented):
Hey @rkzel I ran the code and it's working well. I also think the logic you implemented makes sense, we just need to figure out if this is the best dependency to handle "clickouts". @chadoh what do you think?

Also, if you update the dependencies, please make sure the package.json is updated as well.

Schwartz10 reviewed (dismissed):
Hey @rkzel let's implement this with chad's strategy and not add any new deps

rkzel committed:
Merge branch 'master' of github.com:AutarkLabs/aragon-profile into 34-upload-image

rkzel committed:
replaced react-clickout with simple div

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -102,29 +105,53 @@ const ImageMenu = ({
+      >
+        <div
+          css="padding: 8px 8px 6px 8px"
+          onClick={() => handleToggleMenu(imageTag)}
```
Let's only add `onClick` to `<a>` and `<button>`

chadoh replied:
I think we can get rid of that `handleToggleMenu` definition by changing this line to

    onClick={() => setActive(!active)}

rkzel replied:
Very good point. It started as multi-component thing, so it needed to distinct various active menus - it's no longer the case.

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -28,6 +29,8 @@ const ImageMenu = ({
   imageTitle,
   onSignatures,
 }) => {
+  const [activeImageMenu, toggleImageMenu] = useState('')
```
I don't think we need to name the image menu using the `imageTag` prop. Since each instance of `ImageMenu` on the page will maintain its own internal state, we can use an easier to understand concept like `const [active, setActive] = setState(false)`

rkzel replied:
Done!

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -139,71 +166,61 @@ ImageMenu.propTypes = {
-}
+const getVisibility = props => (isVisible(props) ? 'visible' : 'hidden')
+
+const getBackgroundAlpha = props => (isVisible(props) ? '0.8' : '0')
```
Nice one-liners! Thanks. I think we can clean these functions just a little more by removing the wrapping parentheses. Example:

    const getBackgroundAlpha = props => isVisible(props) ? '0.8' : '0'

rkzel replied:
Yes, but eslint/prettier has different opinion. Actually it's the second time eslint has problem with something that can be allowed, but `prettier` decides to be strict about it. What do we do in such cases? I tried adding:
`"no-extra-parens": [2, "all"]`
but it conflicts with prettier and both cases are highlighted, with and without parentheses. I found no quick way of telling prettier to acknowledge that if there is no problem, there is really no problem.
I'm going to leave it for now.

chadoh replied:
Oh, I didn't know we had prettier configured here. In that case, I'll abide by the rule, "no code style discussion, ever. add linter rule or stfu." 

🤐

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -139,71 +166,61 @@ ImageMenu.propTypes = {
+    border: 1px solid rgba(209, 209, 209, 0.8);
   }
+
+  visibility: ${props => getVisibility(props)};
```
No need for the intermediate function declaration. This works the same:

    visibility: ${getVisibility};

rkzel replied:
Oh, I didn't know - it works, thank you!

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -139,71 +166,61 @@ ImageMenu.propTypes = {
+
+  background-color: rgba(255, 255, 255, ${props => getBackgroundAlpha(props)});
+  color: rgba(0, 0, 0, ${props => getBackgroundAlpha(props)});
   border: ${props => getBorder(props)};
```
Same here

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -139,71 +166,61 @@ ImageMenu.propTypes = {
-    transition: visibility 0.3s linear, opacity 0.3s linear;
-    visibility: visible;
-    opacity: 0.8;
+    transition: all 0.3s linear;
```
I see we're adding this here, to `.imageHover:hover`, and we're adding it to the root `ImageMenuStyled` on line 198. Do we need both?

rkzel replied:
Forsooth. Removed the surplus one.

chadoh commented on src/client/components/Profile.js:
```diff
@@ -9,11 +9,7 @@ import WorkHistoryPanel from './WorkHistoryPanel'
-      width: 100%;
-    `}
-  >
+  <div css="width: 100%">
```
🎉

rkzel replied:
:)

chadoh commented on src/client/components/styled-components/index.js:
```diff
@@ -112,6 +112,8 @@ export { default as IconLocation } from './IconLocation'
 export { default as AnimationLoadingCircle } from './AnimationLoadingCircle'
 export { default as IconSuccess } from './IconSuccess'
 export { default as IconError } from './IconError'
+
```
let's get rid of this commit noise

rkzel replied:
Sure.

chadoh commented on src/client/components/styled-components/IconCamera.js:
```diff
@@ -0,0 +1,47 @@
+}
+
+IconCamera.defaultProps = {
+  color: '#000',
```
Is `color` a meaningful attribute, on an `svg` element? 

Regardless, it looks like we are not using it, so I'd rather we didn't add it to this file.

rkzel replied:
Rather than removing it, I changed hardcoded color to:
`fill={props.color}`
So when we need it, it's there. We do similar things with other icons, so they should all have configurable width, height and color.

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -139,71 +166,61 @@ ImageMenu.propTypes = {
-  }
-  > :not(:last-child) {
-    margin-bottom: 1px;
+  > :not(:first-child) :hover {
```
Let's nest these inside the `> :not(:first-child)` block below

rkzel replied:
Sure.

rkzel committed:
post-review fixes

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -103,28 +105,43 @@ const ImageMenu = ({
+          imageCid,
+        })}
+      >
+        <div css="padding: 8px 8px 6px 8px" onClick={() => setActive(!active)}>
```
Let's only add `onClick` to `<a>` or `<button>`, please

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -103,28 +105,43 @@ const ImageMenu = ({
+        {active && (
+          <React.Fragment>
+            <input {...getInputProps({ disabled: false })} />
+            <div css="padding: 8px 12px 6px 12px" onClick={open}>
```
Same here

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -103,28 +105,43 @@ const ImageMenu = ({
+            {imageExists && (
+              <div
+                css="padding: 6px 12px 8px 12px"
+                onClick={() => dispatchModal(removeItem(imageCid, imageTag))}
```
and here

chadoh commented on src/client/components/styled-components/IconCamera.js:
```diff
@@ -0,0 +1,47 @@
+import PropTypes from 'prop-types'
+
+const IconCamera = props => (
+  <svg width="22px" height="18px" viewBox="0 0 22 18" {...props}>
```
The `{...props}` will add a `color={props.color}` here. I think the component might complain if we ever start using that.

On the line above, maybe we can destructure the props, `IconCamera = ({ height, width, color }) => (`

chadoh committed:
ImageMenu: accessibility & UX tweaks

Accessibility fixes:

* use buttons, not divs, for everything with a click handler
* don't use `visibility="hidden"`, since this hides the buttons from
  screen readers
* make the dropdown a _live region_, controlled by the button. Screen
  readers watch for chages in this region, and announce them politely

While getting there, some UX tweaks were needed:

* rework the styling to account for the buttons
* simplify the styling for a smoother effect that more closely matches
  Soundcloud's
* fix the click-away handling, creating a new component that doesn't
  fire onBlur when the click happens inside its own container. This
  `ClickOutHandler` lives right within the `ImageMenu` file for now.
* don't bother showing the "more actions" dropdown if there's no image
  yet, since the only thing to be done is upload a new image. This again
  matches Soundcloud's design.

chadoh commented on src/client/components/ImageMenu.js:
```diff
@@ -103,28 +105,43 @@ const ImageMenu = ({
-    </ImageMenuStyled>
+
+        {active && (
+          <React.Fragment>
```
While I'm thinking about accessibility, another interaction I anticipate being strange for users of screen readers:

Once "Update Cover photo" is a button, and therefore findable by screen readers, here's what will happen:

* I click "Update Cover photo"
* ...

Nothing. I don't know that new stuff appeared on the screen. I would have to start reading through the page again or something to discover that there are two new buttons available.

I think we can fix that by using an `aria-live` region. Here's how we implement it:

```jsx
<button onClick={() => setActive(true)} aria-controls={`${imageTag}-photo-actions`}>
  Update {imageTitle} photo
</button>
<div role="region" aria-live="polite" id={`${imageTag}-photo-actions`}>
  {active && (
    <React.Fragment>...</React.Fragment>
  )}
</div>
```

Here's how it works:

* on initial page load, the screen reader makes note of this "aria-live" element
* when the content in this region changes, the screen reader breaks the flow of whatever it's doing and reads the content

Learn more here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions

chadoh reviewed (approved):
See some obsolete comments below – I just updated this branch with the changes I wanted to see. It took a fair bit of finagling to get all of the interactions working properly. Here's how it looks like now, for various modes of interaction:

## Adding or updating an image with your mouse

Notice that there's no dropdown if there's no image, since "delete" makes no sense in this scenario.

![profile adding or updating image with mouse](https://user-images.githubusercontent.com/221614/61079543-1833bb00-a3f1-11e9-843f-0271d1d5771a.gif)

## Dragging in an image

![profile drag and drop](https://user-images.githubusercontent.com/221614/61079544-1833bb00-a3f1-11e9-85d0-fb12ca069d20.gif)

## Tabbing through the page

I'm tabbing backward through the page in this gif, using shift-Tab on my keyboard.

![profile keyboard navigation](https://user-images.githubusercontent.com/221614/61079545-1833bb00-a3f1-11e9-8948-c138a267e3cd.gif)

## Removing images using your mouse

![profile removing images with mouse](https://user-images.githubusercontent.com/221614/61079547-18cc5180-a3f1-11e9-99ac-c0e00ac98bf4.gif)

## Using a screen reader

The screen reader speaks out the stuff you see written at the bottom. Notice that, after clicking the "Update Cover photo" button, it speaks out what's in the dropdown. "Upload new image Delete" – this is what the `role="region" aria-live="polite"` stuff gives us. The "Form Controls" menu that you see appear is a screen reader feature that lists all form controls on the page, which can be summoned with control-option-U on a Mac.

![profile screen reader behavior](https://user-images.githubusercontent.com/221614/61079548-18cc5180-a3f1-11e9-823e-539792d65382.gif)
